src/core/Table/hooks/useColumnVisibility.tsx

```diff
@@ -0,0 +1,71 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { ColumnInstance, HeaderProps, Hooks } from 'react-table';
+import { Checkbox } from '../..';
+import SvgColumnManager from '@itwin/itwinui-icons-react/cjs/icons/ColumnManager';
+import { DropdownMenu } from '../../DropdownMenu';
+import { IconButton } from '../../Buttons/IconButton';
+import { MenuItem } from '../../Menu';
+
+export const COLUMN_MANAGER_ID = 'iui-table-column-manager';
+
+export const useColumnVisibility = <T extends Record<string, unknown>>(
+  hasColumnManager: boolean,
+) => (hooks: Hooks<T>) => {
+  if (!hasColumnManager) {
+    return;
+  }
+
+  hooks.allColumns.push((columns: ColumnInstance<T>[]) => [
+    ...columns,
+    {
+      id: COLUMN_MANAGER_ID,
+      minWidth: 48,
+      width: 48,
+      maxWidth: 48,
+      disableReordering: true,
+      columnClassName: 'iui-slot',
+      cellClassName: 'iui-slot',
+      Header: ({ allColumns }: HeaderProps<T>) => {
+        const [isOpen, setIsOpen] = React.useState(false);
+        const headerCheckBoxes = () =>
+          allColumns
+            .filter(({ id }) => !id.includes('iui-table'))
+            .map((column) => {
+              const { onChange, checked } = column.getToggleHiddenProps();
+              return (
+                <MenuItem
+                  key={column.id}
+                  {...column.getToggleHiddenProps()}
+                  icon={
+                    <Checkbox
+                      id={`iui-column-${column.id}`}
+                      checked={checked}
+                      disabled={column.disableToggleVisibility}
+                    />
+                  }
+                  onClick={() => onChange({ target: { checked: !checked } })}
+                  disabled={column.disableToggleVisibility}
+                >
+                  {column.Header}
+                </MenuItem>
+              );
+            });
+        return (
+          <DropdownMenu
+            menuItems={headerCheckBoxes}
```

Comment:
Another thing.. Would there be a case that users want a little different content there? Maybe add additional description or something..
Might be worth thinking, if we can provide also a way to override our default behavior

Reply:
@bentleyvk any ideas on this one?

Reply:
Maybe for now we can release as it is and if someone asks for such feature we could add some property to the column like `columnName` or smth